Feature/payment-component

camposjoaoc opened the pull request:
This PR introduces the `PaymentForm` component, responsible for handling the credit card payment flow.

### Highlights:
- Implementation of the `PaymentForm`;
- Includes field-level validations for:
  - Cardholder name
  - Card number with formatting (#### #### #### ####)
  - Expiration month and year
  - CVC
- State is managed through a dedicated Zustand store (`paymentStore.ts`) (let me know if I messed up anything with the Zustand part).
- Strong typing enforced via `PaymentFormTypes` and `ValidationRule` types.
- Controlled and masked input for numeric-only fields.
- Submit button triggers validation and prints data to the console.

### UI:
- Klarna and Swish buttons replace Apple Pay and PayPal.
- Introduced `.button-secondary__payment` class in `index.css` for payment-style buttons (white background, black border.

This component is prepared for future integration with the backend, but the handleSubmit function still needs to be tested once the backend route is ready.

Note: I was testing the component by importing it into the Shop page (since that page is currently empty) and, I think my prettier change the indentation on index.css, sorry 😥

reevcrona commented on frontend/vite-project/src/types/paymentFormTypes.ts:
```diff
@@ -0,0 +1,25 @@
+    (value: string) => boolean,
+    string
+];
+
```
```suggestion

export interface PaymentActions { 
    setFormData: (data: PaymentFormTypes) => void;
    updateFormField: <K extends keyof PaymentFormTypes>(
        key: K,
        value: PaymentFormTypes[K]
    ) => void;
  }
```
Separate functions that modify state into an actions object.
Combining the actions type with the PaymentState type.

reevcrona commented on frontend/vite-project/src/types/paymentFormTypes.ts:
```diff
@@ -0,0 +1,25 @@
+    string
+];
+
+export interface PaymentState { 
+    formData: PaymentFormTypes;
+    setFormData: (data: PaymentFormTypes) => void;
+    updateFormField: <K extends keyof PaymentFormTypes>(
+        key: K,
+        value: PaymentFormTypes[K]
+    ) => void;
+  }
```
```suggestion
export interface PaymentState { 
    formData: PaymentFormTypes;
    actions: PaymentActions
  }
```

reevcrona commented on frontend/vite-project/src/stores/paymentStore.ts:
```diff
@@ -0,0 +1,21 @@
+        cvc: '',
+        saveCard: false,
+    },
+    setFormData: (data) => set({ formData: data }),
+    updateFormField: (key, value) =>
+        set((state) => ({
+            formData: {
+                ...state.formData,
+                [key]: value,
+            },
+        })),
```
```suggestion
    actions:{
    setFormData: (data) => set({ formData: data }),
    updateFormField: (key, value) =>
        set((state) => ({
            formData: {
                ...state.formData,
                [key]: value,
            },
        })),
    }
    
```

reevcrona replied:
```suggestion
        })),
 export const useFormData = () => usePaymentStore((state) => state.formData);
 export const usePaymentActions = () => usePaymentStore((state) => state.actions)
```
Export custom hooks for cleaner and more modular access to the shipping store
useFormData returns the current form data state.
usePaymentActions provides access to all state-modifying actions.

reevcrona commented on frontend/vite-project/src/stores/paymentStore.ts:
```diff
@@ -0,0 +1,21 @@
+import { create } from 'zustand';
+import { PaymentFormTypes, PaymentState } from '../types/paymentFormTypes';
+
+export const usePaymentStore = create<PaymentState>((set) => ({
```
```suggestion
const usePaymentStore = create<PaymentState>((set) => ({
```

reevcrona commented on frontend/vite-project/src/components/payment-form/PaymentForm.tsx:
```diff
@@ -0,0 +1,179 @@
+import { useState } from "react";
+import { usePaymentStore } from "../../stores/paymentStore";
```
```suggestion
import { useFormData,usePaymentActions } from "../../stores/paymentStore";
```

reevcrona commented on frontend/vite-project/src/components/payment-form/PaymentForm.tsx:
```diff
@@ -0,0 +1,179 @@
+import "../payment-form/PaymentForm.css";
+
+const PaymentForm = () => {
+    const { formData, updateFormField } = usePaymentStore();
```
```suggestion
   const formData= useFormData();
   const {updateFormField} = usePaymentActions();
```
I'm sure you did this before I realized my mistake. Leaving it as it was would probably be fine for now, but if we added another piece of state, it could lead to issues. I just want to encourage good habits by using custom hooks for accessing state.

reevcrona reviewed (commented):
Make sure every input has an associated label, this is important for accessibility and overall form usability. 

You have lots of great validation functions! You could consider moving them into a custom hook like useFormValidation() to separate the UI from the logic. Who knows this might be useful in other forms later on, and we could simply import the validation functions we need.

That said, this is not a priority at all. Great work, Joao, the page looks great and the validation works. Awesome job!

camposjoaoc commented:
Hi @reevcrona, I changed the code according to your suggestions. I agree that we can move the validations into a custom hook but, since is just for the credit card, we can add this idea on the nice-to-have list. Also, I added labels but using this class `sr-only` to keep the label hidden and according to the wire-frame, this is a way of making **visually hidden but screen reader accessible**. I made a test with the Mac voice assistant and works perfectly well. 

Thanks for the suggestions and the feedback 😊. Feel free to review and if is everything ok, we can merge.

reevcrona commented:
> Hi @reevcrona, I changed the code according to your suggestions. I agree that we can move the validations into a custom hook but, since is just for the credit card, we can add this idea on the nice-to-have list. Also, I added labels but using this class `sr-only` to keep the label hidden and according to the wire-frame, this is a way of making **visually hidden but screen reader accessible**. I made a test with the Mac voice assistant and works perfectly well.
> 
> Thanks for the suggestions and the feedback 😊. Feel free to review and if is everything ok, we can merge.

reevcrona commented:
Clicked the wrong button lol. 'I added labels but using this class sr-only' very good i will add this to my forms later.
